### PR TITLE
inital port of ConcatDataset

### PR DESCRIPTION
**Pull request by esc**
Starting to work on #1759 
Can someone give me feedback before I continue to invest time, please.

a) Is this the right location for this class
b) Where should I add tests? `test_dataloader.py`?
c) Is it still correct to remove the `numpy` dependency?

### CODE REVIEW

File: torch/utils/data/dataset.py [hunk 1]
```diff
@@ -35,3 +35,51 @@ def __getitem__(self, index):
...
+
+    @staticmethod
+    def cumsum(sequence):
+        result = []
+        sum = 0
+        for s in sequence:
+            l = len(s)
+            cumsum.append(l + sum)
+            sum += l
+        return result
```
Comment (acgtyrant) [inline torch/utils/data/dataset.py]: This function is not necessary, replace it by `sum([len(sequence) for sequence for sequences])`.

File: torch/utils/data/dataset.py [hunk 2]
```diff
@@ -35,3 +35,51 @@ def __getitem__(self, index):
...
+
+        self.datasets = list(datasets)
+        assert len(datasets) > 0, 'datasets should not be an empty iterable'
+        self.cum_sizes = self.cumsum(self.datasets)
+
+    def __len__(self):
+        return self.cum_sizes[-1]
+
+    def __getitem__(self, idx):
+        super(ConcatDataset, self).__getitem__(idx)
```
Comment (colesbury) [inline torch/utils/data/dataset.py]: This line will fail (remove it)

File: torch/utils/data/dataset.py [hunk 3]
```diff
@@ -35,3 +35,51 @@ def __getitem__(self, index):
...
+        dataset_index = self.searchsorted(idx, self.cum_sizes)
```
Comment (colesbury) [inline torch/utils/data/dataset.py]: I think you can just use `bisect` from Python which does binary search:

```
dataset_index = bisect.bisect_left(self.cum_sizes, idx)
```

Raise an `IndexError` if index is out-of-bounds

File: torch/utils/data/dataset.py [hunk 4]
```diff
@@ -35,3 +35,51 @@ def __getitem__(self, index):
...
+
+        if dataset_index == 0:
+            dataset_idx = idx
+        else:
+            dataset_idx = idx - self.cum_sizes[dataset_index - 1]
+
+        return self.datasets[dataset_index][dataset_idx]
```
Comment (colesbury) [inline torch/utils/data/dataset.py]: `dataset_index` vs. `dataset_idx` is confusing. Use something like `sample_idx` for the offset within a dataset

**Review by colesbury — COMMENTED** [review]
You can put the tests in `test_dataloader` or `test_utils`

File: torch/utils/data/dataset.py [hunk 5]
```diff
@@ -35,3 +35,51 @@ def __getitem__(self, index):
...
+    Dataset to concatenate multiple datasets.
+    Purpose: useful to assemble different existing datasets, possibly
+    large-scale datasets as the concatenation operation is done in an
+    on-the-fly manner.
+    Args:
+        datasets (iterable): List of datasets to be concatenated
+    """
+
+    @staticmethod
+    def cumsum(sequence):
```
Comment (fmassa) [inline torch/utils/data/dataset.py]: You can use `torch.cumsum` instead of writing your own.

Comment (esc) [inline torch/utils/data/dataset.py]: `torch.cumsum` seem to operate on tensors, I need something that operates on lists of ints.

File: torch/utils/data/dataset.py (hunk 3 above)
Comment (esc) [inline torch/utils/data/dataset.py]: Yes, almost correct needed `bisect_right`.

**Comment by esc** [comment]
Tests are done. I also found `itertools.accumulate` which was quite nice.

**Comment by esc** [comment]
@colesbury how is pytorchbot implemented?

**Comment by esc** [comment]
Yeah, thought so.

**Comment by esc** [comment]
Have to go to bed now, will fix if still broken tomorrow.

**Comment by esc** [comment]
Travis glitched...

**Comment by esc** [comment]
I don't have perms to restart the build, pushing an empty commit instead.

File: torch/utils/data/dataset.py (hunk 5 above)
Comment (fmassa) [inline torch/utils/data/dataset.py]: You still can do something like `torch.cumsum(torch.LongTensor([1, 2, 3]))`, but I suppose it's not super important in here as it's only used in the constructor.

**Comment by esc** [comment]
Since travis is green

**Comment by esc** [comment]
@colesbury I guess @pytorchbot won't execute my commands, right?

**Review by colesbury — APPROVED** [review]

**Comment by esc** [comment]
Sweet!

**Comment by esc** [comment]
@fmassa anything else you want me to port from `tnt`?

**Pull request merged**